### ISSUE TITLE
Rule proposal: `prefer-error-iserror`

### ISSUE THREAD

**Issue by papb** [sindresorhus/eslint-plugin-unicorn#723]
[Disallow `instanceof Error` checks](https://github.com/sindresorhus/execa/issues/422#issuecomment-625750262).

## Fail

```js
const isError = foo instanceof Error;
```

## Pass

```js
const isError = Object.prototype.toString.call(foo) === '[object Error]'
```

**Comment by fisker** [comment]
How about `no-instanceof`?

**Comment by sindresorhus** [comment]
Too general. It’s generally fine to instanceof your own types, just not built-ins.

**Comment by fisker** [comment]
We have  `Array` checked https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/no-array-instanceof.md

**Comment by sindresorhus** [comment]
Yeah, I think we should just do `Error` for now. `Error` is usually the problematic global.

**Comment by fisker** [comment]
Other built-in errors? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects#Error_objects

**Comment by sindresorhus** [comment]
Yes, the built-in errors too.

**Comment by fregante** [comment]
> Too general. It’s generally fine to instanceof your own types, just not built-ins.

Is it though? If I load the same library in 2 realms I have the same issue. It's just that built-ins are more common.

**Comment by sindresorhus** [comment]
Sounds good

**Comment by papb** [comment]
> 
> 
> Yes, the built-in errors too.

What would be the suggested fix for `foo instanceof TypeError` though?

Since `Object.prototype.toString.call(new TypeError())` gives `'[object Error]'`, I guess it would have to be

```js
Object.prototype.toString.call(foo) === '[object Error]' && foo.name === 'TypeError'
```

instead?

**Comment by sindresorhus** [comment]
Correct

**Comment by tinovyatkin** [comment]
Depending on a project target, but since Node 10.0 `util.types.isNativeError` looks like way cleaner approach than `Object.prototype.toString.call`

**Comment by sindresorhus** [comment]
Cleaner, yes, but less cross-platform. We cannot know where the user intends to run their code.

**Comment by tinovyatkin** [comment]
There is an ESLint environment setting for that.

**Comment by tinovyatkin** [comment]
https://github.com/mysticatea/eslint-plugin-node/blob/master/lib/util/get-configured-node-version.js
https://github.com/amilajack/eslint-plugin-compat/blob/master/src/Versioning.js

**Comment by tinovyatkin** [comment]
Magically smart rule auto fix!

**Comment by sindresorhus** [comment]
@tinovyatkin Sure, if we can reliably detect that it's `node`-only environment, we can suggest that one.

**Comment by sindresorhus** [comment]
This rule is accepted.

**Comment by fisker** [comment]
Suggest hold this for [`Error.isError`](https://github.com/tc39/proposal-is-error)
